Sources/SourceKitLSP/Swift/Rename.swift

```diff
@@ -0,0 +1,446 @@
+///
+/// See `SyntacticRenamePiece` for more details.
+fileprivate struct SyntacticRenameName {
+  let pieces: [SyntacticRenamePiece]
+  let category: SyntacticRenameNameContext
+
+  init?(_ dict: SKDResponseDictionary, in snapshot: DocumentSnapshot, keys: sourcekitd_keys) {
+    guard let ranges: SKDResponseArray = dict[keys.ranges] else {
+      return nil
+    }
+    self.pieces = ranges.compactMap { SyntacticRenamePiece($0, in: snapshot, keys: keys) }
+    guard let categoryUid: sourcekitd_uid_t = dict[keys.category],
+      let category = SyntacticRenameNameContext(categoryUid, keys: keys)
+    else {
+      return nil
+    }
+    self.category = category
+  }
+}
+
+extension SwiftLanguageServer {
+  /// From a list of rename locations, provided by a related identifiers request, compute the list of
+  /// `SyntacticRenameName`s that define which ranges need to be edited to rename a compound decl name.
+  private func getSyntacticRenameRanges(
+    relatedIdentifiers: RelatedIdentifiersResponse,
+    in snapshot: DocumentSnapshot
+  ) async throws -> [SyntacticRenameName] {
+    let locations = SKDRequestArray(sourcekitd: sourcekitd)
+    for relatedIdentifier in relatedIdentifiers.relatedIdentifiers {
+      let position = relatedIdentifier.range.lowerBound
+      guard let utf8Column = snapshot.lineTable.utf8ColumnAt(line: position.line, utf16Column: position.utf16index)
+      else {
+        logger.fault("Unable to find UTF-8 column for \(position.line):\(position.utf16index)")
+        continue
+      }
+      let renameLocation = SKDRequestDictionary(sourcekitd: sourcekitd)
+      renameLocation[keys.line] = position.line + 1
+      renameLocation[keys.column] = utf8Column + 1
+      renameLocation[keys.nameType] = relatedIdentifier.usage.uid(keys: keys)
+      locations.append(renameLocation)
+    }
+    guard let name = relatedIdentifiers.name else {
+      throw ResponseError.unknown("Running sourcekit-lsp with a version of sourcekitd that does not support rename")
+    }
+    let renameLocation = SKDRequestDictionary(sourcekitd: sourcekitd)
+    renameLocation[keys.locations] = locations
+    renameLocation[keys.name] = name
+
+    let renameLocations = SKDRequestArray(sourcekitd: sourcekitd)
+    renameLocations.append(renameLocation)
+
+    let skreq = SKDRequestDictionary(sourcekitd: sourcekitd)
+    skreq[keys.request] = requests.find_syntactic_rename_ranges
+    skreq[keys.sourcefile] = snapshot.uri.pseudoPath
+    // find-syntactic-rename-ranges is a syntactic sourcekitd request that doesn't use the in-memory file snapshot.
+    // We need to send the source text again.
+    skreq[keys.sourcetext] = snapshot.text
+    skreq[keys.renamelocations] = renameLocations
+
+    let syntacticRenameRangesResponse = try await sourcekitd.send(skreq, fileContents: snapshot.text)
+    guard let categorizedRanges: SKDResponseArray = syntacticRenameRangesResponse[keys.categorizedranges] else {
```

Comment:
No, because the C++ portion of `find-syntactic-rename-ranges` is the one that translates results from the Swift `NameMatcher` into ranges. Essentially all of the logic in [SyntacticRenameRangeDetails.cpp](https://github.com/apple/swift/blob/main/lib/Refactoring/SyntacticRenameRangeDetails.cpp) is needed and is not available in SwiftSyntax at the moment.